noether/src/segmentation_server.cpp

```diff
@@ -0,0 +1,168 @@
+#include <ros/ros.h>
+#include <actionlib/server/simple_action_server.h>
+#include <noether_msgs/SegmentAction.h>
+#include <mesh_segmenter/mesh_segmenter.h>
+
+#include <pcl_conversions/pcl_conversions.h>
+#include <pcl/io/vtk_lib_io.h>
+
+
+#include <vtkPolyDataNormals.h>
+#include <vtkCleanPolyData.h>
+#include <vtkWindowedSincPolyDataFilter.h>
+#include <vtkSmoothPolyDataFilter.h>
+
+#include <noether/noether.h>
+#include <vtk_viewer/vtk_utils.h>
+
+class SegmentationAction
+{
+protected:
+  ros::NodeHandle nh_;
+  actionlib::SimpleActionServer<noether_msgs::SegmentAction> server_;
+  std::string action_name_;
+  noether_msgs::SegmentFeedback feedback_;
+  noether_msgs::SegmentResult result_;
+
+public:
+  SegmentationAction(ros::NodeHandle nh, std::string name)
+    : server_(nh_, name, boost::bind(&SegmentationAction::executeCB, this, _1), false), action_name_(name), nh_(nh)
+  {
+    server_.start();
+    ROS_INFO("Segmentation action server online");
+  }
+
+  ~SegmentationAction(void) {}
+
+  void executeCB(const noether_msgs::SegmentGoalConstPtr& goal)
+  {
+    // Step 1: Load parameters
+    // TODO: Pass these in along with action
+    std::string file;
+    double curvature_threshold;
+    int min_cluster_size, max_cluster_size;
+    bool show_individually, save_outputs;
+    nh_.param<std::string>("/mesh_segmenter_client_node/filename", file, "");
+    nh_.param<int>("/mesh_segmenter_client_node/min_cluster_size", min_cluster_size, 500);
+    nh_.param<int>("/mesh_segmenter_client_node/max_cluster_size", max_cluster_size, 1000000);
+    nh_.param<double>("/mesh_segmenter_client_node/curvature_threshold", curvature_threshold, 0.3);
+    nh_.param<bool>("/mesh_segmenter_client_node/show_individually", show_individually, false);
+    nh_.param<bool>("/mesh_segmenter_client_node/save_outputs", save_outputs, false);
+
+    // Convert ROS msg -> PCL Mesh -> VTK Mesh
+    vtkSmartPointer<vtkPolyData> mesh;
+    pcl::PolygonMesh input_pcl_mesh;
+    pcl_conversions::toPCL(goal->input_mesh, input_pcl_mesh);
+    vtk_viewer::pclEncodeMeshAndNormals(input_pcl_mesh, mesh);//, 100, pcl::PointXYZ(0, 50.0, 50.0));
+    vtk_viewer::generateNormals(mesh);
+//    pcl::VTKUtils::convertToVTK(input_pcl_mesh, mesh);            // Converts w
+
+
+
+    // Step 2: Filter the mesh
+    // Create some pointers - not used since passing with VTK pipeline but useful for debugging
+    ROS_INFO("Beginning Filtering.");
+    vtkSmartPointer<vtkPolyData> mesh_in = mesh;
+    vtkSmartPointer<vtkPolyData> mesh_cleaned;
+    vtkSmartPointer<vtkPolyData> mesh_filtered1;
+    vtkSmartPointer<vtkPolyData> mesh_filtered2;
+
+    // Remove duplicate points (Note: this seems to cause problems sometimes)
```

Comment:
remove commented code

Reply:
I left this in to show how to do this in case we need to later. Cleaning the mesh seems like a logical thing to do, and I only disabled it because it seemed to cause some issues. I'm hopeful after I switch to MLS for curvature calculations it will solve some of these problems, and this can be on by default.

@drchrislewis If you'd like me to remove it I can, but that is my rationale

Reply:
Whenever you really want to leave some commented out code in your code. Simply embed your "code" inside some text describing either why you want to keep it, or when and why you'll want to put it back in. Otherwise, it is "not good practice" to leave code commented out. I don't know why, but I've been told this several times, so I repeat the criticism.